File: android/app/src/main/java/com/mwiacek/przepisy/drogowe/SearchProvider.java
package com.mwiacek.przepisy.drogowe;

import android.app.SearchManager;
import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.SharedPreferences;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.net.Uri;
import android.preference.PreferenceManager;
import android.provider.BaseColumns;

import com.fasterxml.jackson.core.JsonFactory;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

public class SearchProvider extends ContentProvider {
    public static final String AUTHORITY = "com.mwiacek.przepisy.drogowe.SearchProvider";

    private static final String[] COLUMN_NAMES = new String[]{
            BaseColumns._ID,
            SearchManager.SUGGEST_COLUMN_TEXT_1,
            SearchManager.SUGGEST_COLUMN_TEXT_2,
            SearchManager.SUGGEST_COLUMN_INTENT_DATA,
            SearchManager.SUGGEST_COLUMN_SHORTCUT_ID};

    private static final int SEARCH_SUGGEST = 1;
    private static final int SHORTCUT_REFRESH = 0;

    private static final UriMatcher uriMatcher = buildUriMatcher();

    private static UriMatcher buildUriMatcher() {
        UriMatcher matcher = new UriMatcher(UriMatcher.NO_MATCH);
        matcher.addURI(AUTHORITY, SearchManager.SUGGEST_URI_PATH_QUERY, SEARCH_SUGGEST);
        matcher.addURI(AUTHORITY, SearchManager.SUGGEST_URI_PATH_QUERY + "/*", SEARCH_SUGGEST);
        matcher.addURI(AUTHORITY, SearchManager.SUGGEST_URI_PATH_SHORTCUT, SHORTCUT_REFRESH);
        matcher.addURI(AUTHORITY, SearchManager.SUGGEST_URI_PATH_SHORTCUT + "/*", SHORTCUT_REFRESH);
        return matcher;
    }

    @Override
    public boolean onCreate() {
        return true;
    }

    @Override
    public Cursor query(Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder) {

        MatrixCursor cursor = new MatrixCursor(COLUMN_NAMES);

        switch (uriMatcher.match(uri)) {
            case SEARCH_SUGGEST:
            case SHORTCUT_REFRESH:
                break;
            //String shortcutId = null;
            //if (uri.getPathSegments().size() > 1) {
            //shortcutId = uri.getLastPathSegment();
            //}
            //return refreshShortcut(shortcutId, projection);
            default:
                throw new IllegalArgumentException("Unknown URL " + uri);
        }

        if (uri.getPath().equals("/search_suggest_query/")) {
            return cursor;
        }

        String searchString = uri.getLastPathSegment().toLowerCase();
        if (searchString.length() == 0) return cursor;

        SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(this.getContext());

        PlikiClass p = new PlikiClass();
        InputStream stream;
        char[] bytes = new char[20000];
        int numRead;
        StringBuilder lines2 = new StringBuilder();
        InputStreamReader inputreader;
        Object[] rowObject;
        String tosearch = "(?![^<]+>)((?i:\\Q" + uri.getLastPathSegment()
                .replace("\\E", "\\E\\\\E\\Q")
                .replace("a", "\\E[aąĄ]\\Q")
                .replace("c", "\\E[cćĆ]\\Q")
                .replace("e", "\\E[eęĘ]\\Q")
                .replace("l", "\\E[lłŁ]\\Q")
                .replace("n", "\\E[nńŃ]\\Q")
                .replace("o", "\\E[oóÓ]\\Q")
                .replace("s", "\\E[sśŚ]\\Q")
                .replace("z", "\\E[zźżŻŹ]\\Q") + "\\E))";

        if (sp.getBoolean("KodyUN_szukanie", false)) {
            boolean Kody = false;
            try {
                stream = getContext().getAssets().open("adr/adr3.htm");
                inputreader = new InputStreamReader(stream);
                lines2.delete(0, lines2.length());

                while ((numRead = inputreader.read(bytes, 0, 20000)) >= 0) {
                    lines2.append(bytes, 0, numRead);
                }

                stream.close();

                if (lines2.toString().replaceFirst(tosearch, "<mark>").contains("<mark>")) {
                    Kody = true;
                }
            } catch (IOException ignore) {
            }
            if (Kody) {
                //counter++;
                rowObject = new Object[]{
                        String.valueOf(3),
                        "Kody UN",
                        "Znaleziono co najmniej raz " + uri.getLastPathSegment(),
                        "u " + uri.getLastPathSegment(),
                        SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT};
                cursor.addRow(rowObject);
            } else {
                //counter++;
                rowObject = new Object[]{
                        String.valueOf(3),
                        "Kody UN",
                        "Nie znaleziono " + uri.getLastPathSegment(),
                        "f " + uri.getLastPathSegment(),
                        SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT};
                cursor.addRow(rowObject);
            }
        }

        if (sp.getBoolean("Znaki_szukanie", true)) {
            boolean Znak = false;
            int l;
            for (int z = 0; z < 15; z++) {
                l = p.znaki[z].length;
                for (int j = 0; j < l; j++) {
                    try {
                        stream = getContext().getAssets().open(p.znaki[z][j] + ".htm");
                        inputreader = new InputStreamReader(stream);
                        lines2.delete(0, lines2.length());
                        while ((numRead = inputreader.read(bytes, 0, 20000)) >= 0) {
                            lines2.append(bytes, 0, numRead);
                        }
                        stream.close();
                        if (!lines2.toString().replaceFirst(tosearch, "<mark>").contains("<mark>"))
                            continue;
                        Znak = true;
                        break;
                    } catch (IOException ignore) {
                    }
                }
                if (Znak) {
                    break;
                }
            }
            if (Znak) {
                //counter++;
                rowObject = new Object[]{
                        String.valueOf(1),
                        "Opisy i nazwy znaków",
                        "Znaleziono co najmniej raz " + uri.getLastPathSegment(),
                        "z " + uri.getLastPathSegment(),
                        SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT};
                cursor.addRow(rowObject);
            } else {
                //counter++;
                rowObject = new Object[]{
                        String.valueOf(1),
                        "Opisy i nazwy znaków",
                        "Nie znaleziono " + uri.getLastPathSegment(),
                        "f " + uri.getLastPathSegment(),
                        SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT};
                cursor.addRow(rowObject);
            }
        }

        if (sp.getBoolean("Kodeks_szukanie", true)) {
            boolean Kodeks = false;
            try {
                stream = getContext().getAssets().open("tekst/kodeks.htm");
                inputreader = new InputStreamReader(stream);
                lines2.delete(0, lines2.length());

                while ((numRead = inputreader.read(bytes, 0, 20000)) >= 0) {
                    lines2.append(bytes, 0, numRead);
                }

                stream.close();

                if (lines2.toString().replaceFirst(tosearch, "<mark>").contains("<mark>")) {
                    Kodeks = true;
                }
            } catch (IOException ignore) {
            }
            if (Kodeks) {
                //counter++;
                rowObject = new Object[]{
                        String.valueOf(2),
                        "Prawo o ruchu drogowym",
                        "Znaleziono co najmniej raz " + uri.getLastPathSegment(),
                        "k " + uri.getLastPathSegment(),
                        SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT};
                cursor.addRow(rowObject);
            } else {
                //counter++;
                rowObject = new Object[]{
                        String.valueOf(2),
                        "Prawo o ruchu drogowym",
                        "Nie znaleziono " + uri.getLastPathSegment(),
                        "f " + uri.getLastPathSegment(),
                        SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT};
                cursor.addRow(rowObject);
            }
        }


        if (sp.getBoolean("Taryfikator_szukanie", true)) {
            boolean taryfikator = false;
            try {
                stream = getContext().getAssets().open("kary/y.jso");
                JsonParser jp = (new JsonFactory()).createParser(stream);

                jp.nextToken();
                jp.nextToken();
                jp.nextToken();
                while (jp.nextToken() != JsonToken.END_ARRAY) {
                    jp.nextToken();
                    jp.nextToken();
                    if (jp.getText().replaceFirst(tosearch, "<mark>").contains("<mark>")) {
                        taryfikator = true;
                        break;
                    }
                    jp.nextToken();
                    jp.nextToken();
                    while (jp.nextToken() != JsonToken.END_ARRAY) {
                        jp.nextToken();
                        jp.nextToken();
                        if (jp.getText().replaceFirst(tosearch, "<mark>").contains("<mark>")) {
                            taryfikator = true;
                            break;
                        }
                        jp.nextToken();
                        jp.nextToken();
                        while (jp.nextToken() != JsonToken.END_ARRAY) {
                            jp.nextToken();
                            jp.nextToken();
                            if (jp.getText().replaceFirst(tosearch, "<mark>").contains("<mark>")) {
                                taryfikator = true;
                                break;
                            }
                            jp.nextToken();
                            jp.nextToken();
                            if (jp.getText().replaceFirst(tosearch, "<mark>").contains("<mark>")) {
                                taryfikator = true;
                                break;
                            }
                            jp.nextToken();
                            jp.nextToken();
                            if (jp.getText().replaceFirst(tosearch, "<mark>").contains("<mark>")) {
                                taryfikator = true;
                                break;
                            }
                            jp.nextToken();

                        }
                        if (taryfikator) {
                            break;
                        }

                        jp.nextToken();
                    }
                    if (taryfikator) {
                        break;
                    }
                    jp.nextToken();
                }

                jp.close();
                stream.close();
            } catch (IOException ignore) {
            }

            if (taryfikator) {
                //counter++;
                rowObject = new Object[]{
                        String.valueOf(4),
                        "Taryfikator od 09.06.2012",
                        "Znaleziono co najmniej raz " + uri.getLastPathSegment(),
                        "t " + uri.getLastPathSegment(),
                        SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT};
                cursor.addRow(rowObject);
            } else {
                //counter++;
                rowObject = new Object[]{
                        String.valueOf(4),
                        "Taryfikator od 09.06.2012",
                        "Nie znaleziono " + uri.getLastPathSegment(),
                        "f " + uri.getLastPathSegment(),
                        SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT};
                cursor.addRow(rowObject);
            }
        }

        if (sp.getBoolean("KierPoj_szukanie", false)) {
            boolean KierPoj = false;
            try {
                stream = getContext().getAssets().open("tekst/kierpoj.htm");
                inputreader = new InputStreamReader(stream);
                lines2.delete(0, lines2.length());

                while ((numRead = inputreader.read(bytes, 0, 20000)) >= 0) {
                    lines2.append(bytes, 0, numRead);
                }

                stream.close();

                if (lines2.toString().replaceFirst(tosearch,
                        "<mark>").contains("<mark>")) {
                    KierPoj = true;
                }
            } catch (IOException ignore) {
            }
            if (KierPoj) {
                //counter++;
                rowObject = new Object[]{
                        String.valueOf(5),
                        "Ustawa o kierujących pojazdami",
                        "Znaleziono co najmniej raz " + uri.getLastPathSegment(),
                        "i " + uri.getLastPathSegment(),
                        SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT};
                cursor.addRow(rowObject);
            } else {
                //counter++;
                rowObject = new Object[]{
                        String.valueOf(5),
                        "Ustawa o kierujących pojazdami",
                        "Nie znaleziono " + uri.getLastPathSegment(),
                        "f " + uri.getLastPathSegment(),
                        SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT};
                cursor.addRow(rowObject);

            }
        }

        return cursor;
    }

    @Override
    public String getType(Uri uri) {
        switch (uriMatcher.match(uri)) {
            case SEARCH_SUGGEST:
                return SearchManager.SUGGEST_MIME_TYPE;
            case SHORTCUT_REFRESH:
                return SearchManager.SHORTCUT_MIME_TYPE;
            default:
                throw new IllegalArgumentException("Unknown URI " + uri);
        }
    }

    @Override
    public Uri insert(Uri uri, ContentValues values) {
        throw new UnsupportedOperationException();
    }

    @Override
    public int update(Uri uri, ContentValues values, String selection, String[] selectionArgs) {
        throw new UnsupportedOperationException();
    }

    @Override
    public int delete(Uri uri, String selection, String[] selectionArgs) {
        throw new UnsupportedOperationException();
    }
}

File: android_kotlin/app/src/main/java/com/mwiacek/przepisy/drogowe/SearchProvider.kt
package com.mwiacek.przepisy.drogowe

import android.app.SearchManager
import android.content.ContentProvider
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.database.MatrixCursor
import android.net.Uri
import android.preference.PreferenceManager
import android.provider.BaseColumns
import com.fasterxml.jackson.core.JsonFactory
import com.fasterxml.jackson.core.JsonToken
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.util.*

class SearchProvider : ContentProvider() {
    override fun onCreate(): Boolean {
        return true
    }

    override fun query(
        uri: Uri,
        projection: Array<String>?,
        selection: String?,
        selectionArgs: Array<String>?,
        sortOrder: String?
    ): Cursor? {
        val cursor = MatrixCursor(COLUMN_NAMES)
        when (uriMatcher.match(uri)) {
            SEARCH_SUGGEST, SHORTCUT_REFRESH -> {
            }
            else -> throw IllegalArgumentException("Unknown URL $uri")
        }
        if (uri.path == "/search_suggest_query/") {
            return cursor
        }
        val searchString = uri.lastPathSegment!!.lowercase(Locale.getDefault())
        if (searchString.length == 0) return cursor
        val sp = PreferenceManager.getDefaultSharedPreferences(this.context)
        val p = PlikiClass()
        var stream: InputStream
        val bytes = CharArray(20000)
        var numRead: Int
        val lines2 = StringBuilder()
        var inputreader: InputStreamReader
        var rowObject: Array<Any>
        val tosearch = "(?![^<]+>)((?i:\\Q" + uri.lastPathSegment!!
            .replace("\\E", "\\E\\\\E\\Q")
            .replace("a", "\\E[aąĄ]\\Q")
            .replace("c", "\\E[cćĆ]\\Q")
            .replace("e", "\\E[eęĘ]\\Q")
            .replace("l", "\\E[lłŁ]\\Q")
            .replace("n", "\\E[nńŃ]\\Q")
            .replace("o", "\\E[oóÓ]\\Q")
            .replace("s", "\\E[sśŚ]\\Q")
            .replace("z", "\\E[zźżŻŹ]\\Q") + "\\E))"
        if (sp.getBoolean("KodyUN_szukanie", false)) {
            var Kody = false
            try {
                stream = context!!.assets.open("adr/adr3.htm")
                inputreader = InputStreamReader(stream)
                lines2.delete(0, lines2.length)
                while (inputreader.read(bytes, 0, 20000).also { numRead = it } >= 0) {
                    lines2.append(bytes, 0, numRead)
                }
                stream.close()
                if (lines2.toString().replaceFirst(tosearch.toRegex(), "<mark>")
                        .contains("<mark>")
                ) {
                    Kody = true
                }
            } catch (ignore: IOException) {
            }
            if (Kody) {
                //counter++;
                rowObject = arrayOf(
                    3.toString(),
                    "Kody UN",
                    "Znaleziono co najmniej raz " + uri.lastPathSegment,
                    "u " + uri.lastPathSegment,
                    SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT
                )
                cursor.addRow(rowObject)
            } else {
                //counter++;
                rowObject = arrayOf(
                    3.toString(),
                    "Kody UN",
                    "Nie znaleziono " + uri.lastPathSegment,
                    "f " + uri.lastPathSegment,
                    SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT
                )
                cursor.addRow(rowObject)
            }
        }
        if (sp.getBoolean("Znaki_szukanie", true)) {
            var Znak = false
            var l: Int
            for (z in 0..14) {
                l = p.znaki[z].size
                for (j in 0 until l) {
                    try {
                        stream = context!!.assets.open(p.znaki[z][j] + ".htm")
                        inputreader = InputStreamReader(stream)
                        lines2.delete(0, lines2.length)
                        while (inputreader.read(bytes, 0, 20000).also { numRead = it } >= 0) {
                            lines2.append(bytes, 0, numRead)
                        }
                        stream.close()
                        if (!lines2.toString().replaceFirst(tosearch.toRegex(), "<mark>")
                                .contains("<mark>")
                        ) continue
                        Znak = true
                        break
                    } catch (ignore: IOException) {
                    }
                }
                if (Znak) {
                    break
                }
            }
            if (Znak) {
                //counter++;
                rowObject = arrayOf(
                    1.toString(),
                    "Opisy i nazwy znaków",
                    "Znaleziono co najmniej raz " + uri.lastPathSegment,
                    "z " + uri.lastPathSegment,
                    SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT
                )
                cursor.addRow(rowObject)
            } else {
                //counter++;
                rowObject = arrayOf(
                    1.toString(),
                    "Opisy i nazwy znaków",
                    "Nie znaleziono " + uri.lastPathSegment,
                    "f " + uri.lastPathSegment,
                    SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT
                )
                cursor.addRow(rowObject)
            }
        }
        if (sp.getBoolean("Kodeks_szukanie", true)) {
            var Kodeks = false
            try {
                stream = context!!.assets.open("tekst/kodeks.htm")
                inputreader = InputStreamReader(stream)
                lines2.delete(0, lines2.length)
                while (inputreader.read(bytes, 0, 20000).also { numRead = it } >= 0) {
                    lines2.append(bytes, 0, numRead)
                }
                stream.close()
                if (lines2.toString().replaceFirst(tosearch.toRegex(), "<mark>")
                        .contains("<mark>")
                ) {
                    Kodeks = true
                }
            } catch (ignore: IOException) {
            }
            if (Kodeks) {
                //counter++;
                rowObject = arrayOf(
                    2.toString(),
                    "Prawo o ruchu drogowym",
                    "Znaleziono co najmniej raz " + uri.lastPathSegment,
                    "k " + uri.lastPathSegment,
                    SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT
                )
                cursor.addRow(rowObject)
            } else {
                //counter++;
                rowObject = arrayOf(
                    2.toString(),
                    "Prawo o ruchu drogowym",
                    "Nie znaleziono " + uri.lastPathSegment,
                    "f " + uri.lastPathSegment,
                    SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT
                )
                cursor.addRow(rowObject)
            }
        }
        if (sp.getBoolean("Taryfikator_szukanie", true)) {
            var taryfikator = false
            try {
                stream = context!!.assets.open("kary/y.jso")
                val jp = JsonFactory().createParser(stream)
                jp.nextToken()
                jp.nextToken()
                jp.nextToken()
                while (jp.nextToken() != JsonToken.END_ARRAY) {
                    jp.nextToken()
                    jp.nextToken()
                    if (jp.text.replaceFirst(tosearch.toRegex(), "<mark>").contains("<mark>")) {
                        taryfikator = true
                        break
                    }
                    jp.nextToken()
                    jp.nextToken()
                    while (jp.nextToken() != JsonToken.END_ARRAY) {
                        jp.nextToken()
                        jp.nextToken()
                        if (jp.text.replaceFirst(tosearch.toRegex(), "<mark>").contains("<mark>")) {
                            taryfikator = true
                            break
                        }
                        jp.nextToken()
                        jp.nextToken()
                        while (jp.nextToken() != JsonToken.END_ARRAY) {
                            jp.nextToken()
                            jp.nextToken()
                            if (jp.text.replaceFirst(tosearch.toRegex(), "<mark>")
                                    .contains("<mark>")
                            ) {
                                taryfikator = true
                                break
                            }
                            jp.nextToken()
                            jp.nextToken()
                            if (jp.text.replaceFirst(tosearch.toRegex(), "<mark>")
                                    .contains("<mark>")
                            ) {
                                taryfikator = true
                                break
                            }
                            jp.nextToken()
                            jp.nextToken()
                            if (jp.text.replaceFirst(tosearch.toRegex(), "<mark>")
                                    .contains("<mark>")
                            ) {
                                taryfikator = true
                                break
                            }
                            jp.nextToken()
                        }
                        if (taryfikator) {
                            break
                        }
                        jp.nextToken()
                    }
                    if (taryfikator) {
                        break
                    }
                    jp.nextToken()
                }
                jp.close()
                stream.close()
            } catch (ignore: IOException) {
            }
            if (taryfikator) {
                //counter++;
                rowObject = arrayOf(
                    4.toString(),
                    "Taryfikator od 09.06.2012",
                    "Znaleziono co najmniej raz " + uri.lastPathSegment,
                    "t " + uri.lastPathSegment,
                    SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT
                )
                cursor.addRow(rowObject)
            } else {
                //counter++;
                rowObject = arrayOf(
                    4.toString(),
                    "Taryfikator od 09.06.2012",
                    "Nie znaleziono " + uri.lastPathSegment,
                    "f " + uri.lastPathSegment,
                    SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT
                )
                cursor.addRow(rowObject)
            }
        }
        if (sp.getBoolean("KierPoj_szukanie", false)) {
            var KierPoj = false
            try {
                stream = context!!.assets.open("tekst/kierpoj.htm")
                inputreader = InputStreamReader(stream)
                lines2.delete(0, lines2.length)
                while (inputreader.read(bytes, 0, 20000).also { numRead = it } >= 0) {
                    lines2.append(bytes, 0, numRead)
                }
                stream.close()
                if (lines2.toString().replaceFirst(
                        tosearch.toRegex(),
                        "<mark>"
                    ).contains("<mark>")
                ) {
                    KierPoj = true
                }
            } catch (ignore: IOException) {
            }
            if (KierPoj) {
                //counter++;
                rowObject = arrayOf(
                    5.toString(),
                    "Ustawa o kierujących pojazdami",
                    "Znaleziono co najmniej raz " + uri.lastPathSegment,
                    "i " + uri.lastPathSegment,
                    SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT
                )
                cursor.addRow(rowObject)
            } else {
                //counter++;
                rowObject = arrayOf(
                    5.toString(),
                    "Ustawa o kierujących pojazdami",
                    "Nie znaleziono " + uri.lastPathSegment,
                    "f " + uri.lastPathSegment,
                    SearchManager.SUGGEST_NEVER_MAKE_SHORTCUT
                )
                cursor.addRow(rowObject)
            }
        }
        return cursor
    }

    override fun getType(uri: Uri): String? {
        return when (uriMatcher.match(uri)) {
            SEARCH_SUGGEST -> SearchManager.SUGGEST_MIME_TYPE
            SHORTCUT_REFRESH -> SearchManager.SHORTCUT_MIME_TYPE
            else -> throw IllegalArgumentException("Unknown URI $uri")
        }
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        throw UnsupportedOperationException()
    }

    override fun update(
        uri: Uri,
        values: ContentValues?,
        selection: String?,
        selectionArgs: Array<String>?
    ): Int {
        throw UnsupportedOperationException()
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        throw UnsupportedOperationException()
    }

    companion object {
        const val AUTHORITY = "com.mwiacek.przepisy.drogowe.SearchProvider"
        private val COLUMN_NAMES = arrayOf(
            BaseColumns._ID,
            SearchManager.SUGGEST_COLUMN_TEXT_1,
            SearchManager.SUGGEST_COLUMN_TEXT_2,
            SearchManager.SUGGEST_COLUMN_INTENT_DATA,
            SearchManager.SUGGEST_COLUMN_SHORTCUT_ID
        )
        private const val SEARCH_SUGGEST = 1
        private const val SHORTCUT_REFRESH = 0
        private val uriMatcher = buildUriMatcher()
        private fun buildUriMatcher(): UriMatcher {
            val matcher = UriMatcher(UriMatcher.NO_MATCH)
            matcher.addURI(AUTHORITY, SearchManager.SUGGEST_URI_PATH_QUERY, SEARCH_SUGGEST)
            matcher.addURI(AUTHORITY, SearchManager.SUGGEST_URI_PATH_QUERY + "/*", SEARCH_SUGGEST)
            matcher.addURI(AUTHORITY, SearchManager.SUGGEST_URI_PATH_SHORTCUT, SHORTCUT_REFRESH)
            matcher.addURI(
                AUTHORITY,
                SearchManager.SUGGEST_URI_PATH_SHORTCUT + "/*",
                SHORTCUT_REFRESH
            )
            return matcher
        }
    }
}